src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -808,6 +827,61 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
       })
     }
 
+    def recordDecl(mods: Modifiers): List[Tree] = {
+      accept(RECORD)
+      val pos = in.currentPos
+      val name = identForType()
+      val tparams = typeParams()
+      val header = formalParams()
+      val superclass = javaLangRecord()
+      val interfaces = interfacesOpt()
+      val (statics, body) = typeBody(RECORD, name)
+
+      // Records generate a canonical constructor and accessors, unless they are manually specified
+      var generateCanonicalCtor = true
+      var generateAccessors = header
+        .view
+        .map { case ValDef(_, name, tpt, _) => name -> tpt }
+        .toMap
+      for (DefDef(_, name, List(), List(params), tpt, _) <- body) {
+        if (name == nme.CONSTRUCTOR && params.size == header.size) {
+          val ctorParamsAreCanonical = params.lazyZip(header).forall {
+            case (ValDef(_, _, tpt1, _), ValDef(_, _, tpt2, _)) => tpt1 equalsStructure tpt2
```

Comment:
Is `equalsStructure` the right way to check that the method signatures match?

Reply:
Right. This is exactly what I was worried about. Thank you for the pointer to similar code.

I'm a little concerned that the perfect solution won't be possible, but I'll try. If it isn't, I think we should lean towards assuming the constructor is different. It would be undesirable to assume the constructor is the same in a case like the following:

```java
public record Test(String s) {
  public Test(some.other.fully.qualified.String s) {
    this.s = s;
  }
}
```

Unlike the case where we incorrectly mark `String` and `java.lang.String` as different, incorrectly assuming `String` and `some.other.fully.qualified.String` are the same isn't fixable from the Java source without renaming `some.other.fully.qualified.String` (something which might not even be possible if it comes from another package).